File: lib/tools/bin/makeManifest.pl
#!/usr/local/bin/perl
#
# This script traverses the Gallery tree and creates a manifest file that
# contains a list of checksums for files in the distribution.  The installer
# verifies the integrity of the files before proceeding.
#
use strict;
use File::Basename;
use Cwd;
use String::CRC32;
my $quiet = 0;
foreach my $i (0 .. $#ARGV) {
    if ('-q' eq $ARGV[$i]) {
	$quiet = 1;
    }
}
sub quietprint {
    if (!$quiet) {
	my $string = shift;
	print STDERR $string;
    }
}
$| = 1;

chdir(dirname($0) . '/../../..');
my $basedir = cwd();

# Get a list of every file committed to Subversion.
#
my @entries = ();
quietprint("Finding all files...");
&listSvn(\@entries);
quietprint("\n");

# Strip base dir, sort
quietprint("Sorting...");
@entries = sort @entries;
quietprint("\n");

# Split into sections
#
my %sections;
quietprint("Separating into sections...");
foreach my $file (@entries) {
  if ($file =~ m{^((modules|layouts|themes)/.*?)/}) {
    push(@{$sections{"$1/MANIFEST"}}, $file);
  } else {
    push(@{$sections{'MANIFEST'}}, $file);
  }
}
quietprint("\n");

# Now generate the checksum files
#
quietprint("Generating checksums...");
my $changed = 0;
my $total = 0;
foreach my $manifest (keys %sections) {
  my @old_lines = ();
  my $oldContent = '';
  my $oldRevision = '';
  if (open(FD, "<$manifest")) {
    @old_lines = <FD>;
    close(FD);
    $oldRevision = $1 if ($old_lines[0] =~ /Revision(: \d+\s*)\$/);
    $oldContent = join('', @old_lines);
  }
  open(my $out, ">$manifest.new") or die;
  print $out '# $Revision' . "$oldRevision\$\n";
  print $out "# File crc32 crc32(crlf) size size(crlf)  or  R File\n";
  my @entries = @{$sections{$manifest}};
  my %deleted;
  my %seen = {};
  foreach my $entry (@entries) {
    my ($file, $isBinary) = split(/\@\@/, $entry);
    next if ($file =~ /MANIFEST$/);

    if ($file =~ /deleted:(.*)/) {
      $deleted{$1}++;
    } else {
      $seen{$file}++;
      open(my $fd, "<$file");
      binmode($fd);
      my $data = join('', <$fd>);
      close($fd);

      my ($data_crlf, $size, $size_crlf);
      if ($isBinary) {
	$data_crlf = $data;
	$size = $size_crlf = (stat($file))[7];
      } else {
	if ($data =~ /\r/) {
	  ($data_crlf = $data) =~ s/\r//g;
	} else {
	  ($data_crlf = $data) =~ s/\n/\r\n/g;
	}
	$size = length($data);
	$size_crlf = length($data_crlf);
      }

      my $cksum = crc32($data);
      my $cksum_crlf = crc32($data_crlf);
      print $out "$file\t$cksum\t$cksum_crlf\t$size\t$size_crlf\n";
    }
  }
  if (@old_lines) {
    foreach (@old_lines) {
      next if /^\#/;
      if (/^R\t(.*)$/) {
	if ($seen{$1}) {
	  # Marked as deleted in the old manifest but back again!
	} else {
	  $deleted{$1}++;
	}
      } else {
	/^(\S+)\t/;
	unless ($seen{$1}) {
	  # We used to have it but we don't anymore
	  $deleted{$1}++;
	}
      }
    }

    print $out map("R\t$_\n", sort rSort keys %deleted);
  }
  close($out);

  $changed += replaceIfNecessary($oldContent, $manifest, "$manifest.new");
  $total++;

  quietprint(".");
}
quietprint("\n");
quietprint(sprintf("Completed in %d seconds\n", time - $^T));
quietprint(sprintf("Manifests changed: $changed (total: $total)\n"));

sub replaceIfNecessary {
  my ($oldContent, $oldFile, $newFile) = @_;

  open(FD, "<$newFile") || die;
  my $new = join("", <FD>);
  close(FD);

  if ($oldContent ne $new) {
    rename($newFile, $oldFile);
    return 1;
  } else {
    unlink($newFile);
    return 0;
  }
}

sub listSvn {
  my $entries = shift;
  my %binaryList = ();
  local *FD;
  open(FD, "svn propget --non-interactive -R svn:mime-type |") or die;
  while (<FD>) {
    split / - /;
    $binaryList{$_[0]} = 1;
  }
  close FD;
  open(FD, "svn status --non-interactive -v -q |") or die;
  while (<FD>) {
    die "\n$_" unless /^(.).....\s*\d+\s+\d+\s+\S+\s+(.*)$/;
    die "\n$2" unless (-e $2);
    next unless (-f $2);
    die "Check $1 status for $2" if ($1 ne ' ' and $1 ne 'D' and $1 ne 'M');
    quietprint("Warning: $2 is locally modified\n") if ($1 eq 'M');
    push(@$entries,
      sprintf("%s%s@@%d", ($1 eq 'D' ? 'deleted:' : ''), $2, exists($binaryList{$2})));
  }
  close FD;
}

sub rSort {
  return 1 if ("$a/" eq substr($b, 0, length($a) + 1));
  return -1 if ("$b/" eq substr($a, 0, length($b) + 1));
  return $a cmp $b;
}

File: lib/tools/bin/makeManifest.php
<?php
/*
 * Gallery - a web based photo album viewer and editor
 * Copyright (C) 2000-2007 Bharat Mediratta
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or (at
 * your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street - Fifth Floor, Boston, MA  02110-1301, USA.
 */
if (!function_exists('file_put_contents')) {
    /* Define file_put_contents if running PHP 4.x */
    function file_put_contents($path, $data) {
	if (false === ($file = fopen($path, 'w')) || false === fwrite($file, $data)) {
	    return false;
	}
	fclose($file);
	return true;
    }
}
    
function makeManifest() {
    global $SRCDIR;
    $startTime = time();
    
    /* Just so we are consistent lets standardize on Unix path sepearators */
    $baseDir = $SRCDIR . '/gallery2/';
    chdir($baseDir);
    
    quiet_print("Finding all files...");
    $entries = listSvn($baseDir);

    quiet_print("Sorting...");
    sort($entries);

    /* Split into sections */
    $sections = array();
    quiet_print("Separating into sections...");
    foreach ($entries as $file) {
	$matches = array();
	if (preg_match('#(^.*)((modules|layouts|themes)[\\/].*?)[\\/]#', $file, $matches) !== 0) {
	    $sections["{$matches[2]}/MANIFEST"][] = $file;
	} else {
	    $sections['MANIFEST'][] = $file;
	}
    }
    
    /* Now generate the checksum files */
    quiet_print("Generating checksums...");
    $changed = 0;
    $total = 0;
    
    foreach ($sections as $manifest => $entries) {
	$oldLines = file($manifest);
	$oldContent = implode('', $oldLines);
	$nl = preg_match('/\r\n/', $oldContent) ? "\r\n" : "\n";
	$matches = array();
	preg_match('/Revision: (\d+\s*)\$/', $oldLines[0], $matches);
	$oldRevision = $matches[1];
    
	$newContent = "# \$Revi" . "sion: $oldRevision\$$nl";
	$newContent .= "# File crc32 crc32(crlf) size size(crlf)  or  R File$nl";
    
	$deleted = $seen = array();
	foreach ($entries as $entry) {
	    list ($file, $isBinary) = preg_split('/\@\@/', $entry);
	    
	    if (preg_match('/deleted:(.*)/', $file, $matches)) {
		$deleted[$matches[1]] = true;
	    } else {
		$seen[$file] = true;
		$fileHandle = fopen($file, 'rb');
		$fileSize = filesize($file);
		$data = fread($fileHandle, $fileSize);
		fclose($fileHandle);
    
		$data_crlf = $data;
		if ($isBinary) {
		    $size = $size_crlf = filesize($file);
		} else {
		    if (preg_match("/\r\n/", $data)) {
			$data = str_replace("\r\n", "\n", $data);
		    } else {
			$data_crlf = str_replace("\n", "\r\n", $data_crlf);
		    }
		    $size = strlen($data);
		    $size_crlf = strlen($data_crlf);
		}
		
		$cksum = crc32($data);
		$cksum_crlf = crc32($data_crlf);
		$file = substr($file, strlen($baseDir));
		$newContent .= 
		    sprintf("$file\t%u\t%u\t%d\t%d$nl", $cksum, $cksum_crlf, $size, $size_crlf);
	    }
	}
    
	if (!empty($oldLines)) {
	    foreach ($oldLines as $line) {
		if ($line[0] == '#') {
		    continue;
		}
		if (preg_match('/^R\t(.*)$/', $line, $matches)) {
		    $file = $baseDir . trim($matches[1]);
		    if (empty($seen[$file])) {
			$deleted[$file] = true;
		    }
		} else {
		    preg_match('/^(.+?)\t/', $line, $matches);
		    $file = $baseDir . trim($matches[1]);
		    if (empty($seen[$file])) {
			$deleted[$file] = true;
		    }
		}
	    }
    
	    foreach ($deleted as $file => $unused) {
		$file = substr($file, strlen($baseDir));
		$newContent .= "R\t$file$nl";
	    }
	}
    
	if ($oldContent != $newContent) {
	    file_put_contents($manifest, $newContent);
	    $changed++;
	}
	$total++;    
    }
    
    quiet_print(sprintf("Completed in %d seconds", time() - $startTime));
    quiet_print(sprintf("Manifests changed: $changed (total: $total)"));
}

/**
 * Retrieve the SVN Entries
 * @param string $path Path to create retrieve the SVN entries for.
 * @return array List of SVN entries
 */
function listSvn($path) {
    $entries = array();

    $binaryList = array();
    exec("svn propget --non-interactive -R svn:mime-type $path", $output);
    foreach ($output as $line) {
	$parts = preg_split('/\s-\s/', $line);
	$file = str_replace("\\", '/', $parts[0]);
	$binaryList[$file] = 1;
    }

    $output = array();
    exec("svn status --non-interactive -v -q $path", $output);
    foreach ($output as $line) {
	$matches = array();
	if (preg_match('/^(.).....\s*\d+\s+[\d|\?]+\s+\S+\s+(.*)$/', $line, $matches) == 0) {
	    die("Unexpected SVN status format:\n$line\n");
	}
	if (!file_exists($matches[2])) {
	    die("The file '$matches[2]' does not exist");
	}

	if (is_dir($matches[2])) {
	    continue;
	}

	if (preg_match('#[\\/]MANIFEST#', $matches[2]) > 0) {
	    continue;
	}

	if ($matches[1] == 'M') {
	    quiet_print("Warning: $matches[2] is locally modified");
	} else if (!in_array($matches[1], array(' ', 'D', 'M'))) {
	    die("Check {$matches[1]} status for {$matches[2]}");
	}

	$status = $matches[1] === 'D' ? 'deleted:' : '';

	$file = str_replace("\\", '/', $matches[2]);
	$entries[] = sprintf("%s%s@@%d", $status, $file, isset($binaryList[$file]));
    }
    return $entries;
}

?>
